Widgets/Deployment/WidgetsKvCreation.Ps1

```diff
@@ -33,54 +33,129 @@ function Get-SentinelApplicationObjectId {
+            if((Get-AzRoleAssignment  -RoleDefinitionName "Contributor" -SignInName $userName -ExpandPrincipalGroups| Where-Object {$_.Scope -eq "/subscriptions/$SubscriptionId"}).Count -gt 0) {
+                $cloudShellProvider = Register-AzResourceProvider -ProviderNamespace Microsoft.CloudShell | out-null
+                $keyValutProvider = Register-AzResourceProvider -ProviderNamespace Microsoft.KeyVault | out-null
+                Write-Log "Initiated CloudShell and KeyVault registreation. This might take some time. If the KeyVault deployment fails, try running the script again in a few minutes "
+            } else {  # user doesn't have permissions to perform the registrations
+                Write-Log "Resource Provider Registrations are missing and the user lacks Contributor Role over the subscription to perform the registrations. Deployment may fail until registrations are performed. Contant SentinelWidgets@microsoft.com for support" -isError
+            }
+        } 
+    } catch {
+        Write-Log "TryRegisterProviders Filed with exception: $_" -isError
+    }
 }
 
 ##############################################################################################
 ####  Main
 ##############################################################################################
 $sub = Select-AzSubscription $SubscriptionId
-Register-AzResourceProvider -ProviderNamespace Microsoft.CloudShell | out-null
+Write-Log "Selected a subscription. Name - '$($($sub.Subscription).Name)'; Id - '$($($sub.Subscription).Id)'"
+$userName = az account show --query user.name --output tsv
+
+TryRegisterProviders $userName
 
-"Selected a subscription. Name - '$($($sub.Subscription).Name)'; Id - '$($($sub.Subscription).Id)'"
 $workspace = Get-AzOperationalInsightsWorkspace | where CustomerId -eq $WorkspaceId
 if ($workspace -eq $null) {
-    Print-Error "Failed to find a workspace with id '$WorkspaceId'. Fix the WorkspaceId or contant SentinelWidgets@microsoft.com for support. Aborting."
+    Write-Log "Failed to find a workspace with id '$WorkspaceId'. Fix the WorkspaceId or contant SentinelWidgets@microsoft.com for support. Aborting." -isError
     return
-} 
+}
 
+# Create the KV
 $keyVaultName = Get-WigdetsKeyVaultName -workspaceId $WorkspaceId
 if ((Get-AzKeyVault -Name $keyVaultName) -eq $null) { # KeyVault doesn't exist
-    "KeyValut Deployment in progress, please wait.."
-    $keyVault = New-AzKeyVault -Name $keyVaultName -ResourceGroupName $workspace.ResourceGroupName -Location $workspace.Location -Tag @{WorkspaceId=$WorkspaceId} -WarningAction SilentlyContinue
+    Write-Log "KeyValut Deployment in progress, please wait.."
+
+    $wsResourceGroup = $workspace.ResourceGroupName;
+     if((Get-AzRoleAssignment  -RoleDefinitionName "Contributor" -SignInName $userName -ExpandPrincipalGroups| Where-Object {$_.Scope -eq "/subscriptions/$SubscriptionId/resourceGroups/$wsResourceGroup"}).Count -gt 0) {

```

Comment:
nit:
I see this is used a lot, maybe extract to a function?
the function can have the "scope" (subscription or subscription\resourceGroup) as an argument.

Reply:
Get-AzRoleAssignment appears only twice... I think it will be an over kill. WDYT?